src/core/or/circuitpadding_machines.c

```diff
@@ -0,0 +1,351 @@
+setup_state_machine_for_hiding_hs_circuits(circpad_machine_spec_t *machine)
+{
+  /* Two states: START, BURST (and END) */
+  circpad_machine_states_init(machine, 2);
+
+  /* START -> BURST transition upon first cell received/sent */
+  machine->states[CIRCPAD_STATE_START].
+    next_state[CIRCPAD_EVENT_NONPADDING_RECV] = CIRCPAD_STATE_BURST;
+  machine->states[CIRCPAD_STATE_START].
+    next_state[CIRCPAD_EVENT_NONPADDING_SENT] = CIRCPAD_STATE_BURST;
+
+  /* BURST -> END transition when we finish all the tokens */
+  machine->states[CIRCPAD_STATE_BURST].
+      next_state[CIRCPAD_EVENT_BINS_EMPTY] = CIRCPAD_STATE_END;
+  /* or when the length finishes */
+  machine->states[CIRCPAD_STATE_BURST].
+      next_state[CIRCPAD_EVENT_LENGTH_COUNT] = CIRCPAD_STATE_END;
+
+  /* Keep sending DROP cells without caring of what the other end is doing */
+  machine->states[CIRCPAD_STATE_BURST].
+      next_state[CIRCPAD_EVENT_PADDING_SENT] = CIRCPAD_STATE_BURST;
+  machine->states[CIRCPAD_STATE_BURST].
+      next_state[CIRCPAD_EVENT_PADDING_RECV] = CIRCPAD_STATE_BURST;
+}
+
+/* Setup the BURST state of the machine that hides client-side intro
+ * circuits. */
+static void
+setup_burst_state_for_hiding_intro_circuits(circpad_state_t *burst_state,
+                                            bool is_client)
+{
+  /* Token removal strategy for BURST state */
+  burst_state->token_removal = CIRCPAD_TOKEN_REMOVAL_CLOSEST;
+
+  /* Histogram for BURST state:
+   *
+   * We want a machine that will send up to 20 fake cells over a reasonable
+   * time frame, to obfuscate the fact that introduction circuits always send 4
+   * cells both ways */
+
+  /* Figure out the length of the BURST state so that it's randomized. See
+     definition of these constants for rationale. */
+  burst_state->length_dist.type = CIRCPAD_DIST_UNIFORM;
+  burst_state->length_dist.param1 = is_client ?
+    HS_MACHINE_PADDING_MINIMUM_CLIENT : HS_MACHINE_PADDING_MINIMUM_SERVICE;
+  burst_state->length_dist.param2 = is_client ?
+    HS_MACHINE_PADDING_MAXIMUM_CLIENT : HS_MACHINE_PADDING_MAXIMUM_SERVICE;
+
+  /* We mainly care about destroying circuit construction fingerprints. Here is
+   * an example intro circuit construction, where the timestamp granularity is
+   * 1 millisecond:
+   *
+   * IP circuit relay cells:
+   *  0: 11:56:59.703 -> EXTEND2
+   *  1: 11:57:00.074 <- EXTENDED2
+   *  2: 11:57:00.075 -> EXTEND2
+   *  3: 11:57:00.243 <- EXTENDED2
+   *  4: 11:57:17.406 -> EXTEND
+   *  5: 11:57:17.708 <- EXTENDED
+   *  6: 11:57:17.709 -> INTRODUCE1
+   *  7: 11:57:17.904 <- INTRODUCE_ACK
+   *
+   * So with about 200 milliseconds latency between each cell, we appropriately
+   * set the histogram edges to from 30 milliseconds to 100 milliseconds, to
+   * account for faster connections.
+   */
+
+  /* Histogram is: (30 msecs, 200 msecs, infinity). */
+  burst_state->histogram_len = 2;
+  burst_state->histogram_edges[0] = 30000;
+  burst_state->histogram_edges[1] = 200000;
```

Comment:
One problem we'll run in to in research is that this pattern is not going to be in any of their "open worlds" that they train their classifiers against. So even though an argument can be made that this pattern could very well look like an SSH session, or IRC window, or Signal Chat, they won't have that in their testing set and declare this defense broken as a result (because we've just transformed one very unique intro pattern into another "very unique" random pattern that doesn't match anything else, and so their classifiers will still work).

Another option is to make a second state for this machine on the relay side that it transitions to on  receipt of a cell from the client, and have that state send something that looks like a short HTTP response (ie like a handful of packets), before transitioning back to the first state. If we also set use_rtt_estimate=1 on that first state, we should have something that looks like a set of HTTP fetches, which might match something from their open world. I can give an example of this if that doesn't make sense.

Such a machine might be higher overhead, though. It really depends on how long the responses are and how many we do.

What do you think?

Reply:
I definitely agree we are going from a unique pattern to a different unique pattern here and this is a problem. That's why I kinda feel like this is like a poor man's obfs2 for circuit fingerprinting.

![image](https://user-images.githubusercontent.com/3611450/55793949-bf404800-5acc-11e9-9ed7-39ebec2b9568.png)

Just to go a bit deeper into how useful an obfs2 would be here, if we check the backgammon board above, we can see that the top rows of the board are all pretty weirdly looking general circuits. Most of them are either web circuits or directory fetching ones, that don't fit into the usual general pattern (which is the next-to-last one with 48 counts). Perhaps our silly machine could fit into those for now, and try to blend in with the chaos.

--- 
I think the HTTP-simulation could be a plausible way forward for something more sophisticated, but I'm afraid that it's gonna increase the bandwidth, and except if we do it really really methodically, it's still gonna look weird to an observer. I think to make it look semi-legit we would need to impose lots of overhead to every circuit, and also create a proper state machine that simulates HTTP.

I think our main opponent here is the overhead otherwise we could really experiment with such solutions. I did an initial overhead analysis here which left me a bit concerned: https://github.com/torproject/torspec/blob/eb77f96b58fac76a606c5264ac83f68f71902c00/proposals/ideas/xxx-padding-machines-onions.txt#L173
I would be curious to hear what you think, especially since you are much more experienced into thinking about overhead etc.

--- 

Finally, I uploaded the python script I use to make the backgammon board with some instructions. Now you can make your own: https://github.com/asn-d6/padanalyzer

Thanks for the review and the thoughts.

Reply:
So one lightweight option is to use Ticket #30092 (PR #916)  to make some fraction of general circuits look like this crazyness we're going to do with this circuits.

Another option is to implement https://trac.torproject.org/projects/tor/ticket/29203 to allow us to specify some padding machines as reduced padding, and do both the HTTP approach and the janky approach.

A third option is to just implement whatever and monitor padding overhead results, and if they get too high, use our existing padding overhead limits to crank them back down. That might need a fix to https://trac.torproject.org/projects/tor/ticket/29231 though..

I'm starting to sense the need for a bunch of child tickets here.